File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:

        numTable = {}
        hq = []
        for num in nums:
            if num in numTable:
                numTable[num] += 1
            else:
                numTable[num] = 0
        for key, count in numTable.items():
            heapq.heappush(hq, (-count, key))
        
        cnt = 0
        ans = []
        while cnt < k:
            freq, val = heapq.heappop(hq)
            ans.append(val)
            cnt += 1
        return ans

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> freq;
        auto comp = [](const pair<int,int>& a, const pair<int,int>& b) { return a.first > b.first; };

        using MinHeap = priority_queue<pair<int,int>, vector<pair<int,int>>, decltype(comp)>;

        for(auto num : nums) {
            freq[num]++;
        }

        MinHeap minHeap;
        for(auto& [num, count] : freq) {
            minHeap.push({count, num});
            if(minHeap.size() > k) {
                minHeap.pop();
            }
        }

        vector<int> result;
        while(!minHeap.empty()) {
            result.push_back(minHeap.top().second);
            minHeap.pop();
        }
        return result;
    }
};

/*
heap
map <int,int> key, count

priority_queue<
*/